lib/TopStoriesFeed.jsm

```diff
@@ -125,11 +125,26 @@ this.TopStoriesFeed = class TopStoriesFeed {
   }
 
   affinityProividerSwitcher(...args) {
-    let AffinityProvider = UserDomainAffinityProvider;
-    if (this.affinityProviderV2) {
-      AffinityProvider = PersonalityProvider;
+    let affinityProviderV2 = this._prefs.get("affinityProviderV2");
+    if (affinityProviderV2) {
+      try {
+        this.affinityProviderV2 = JSON.parse(affinityProviderV2);
+        if (this.affinityProviderV2 && this.affinityProviderV2.use_v2) {
+          return this.PersonalityProvider(...args, this.affinityProviderV2.model_keys);
+        }
+      } catch (e) {
+        Cu.reportError(`Problem initializing affinity provider v2: ${e.message}`);
+      }
     }
-    return new AffinityProvider(...args);
+    return this.UserDomainAffinityProvider(...args);
+  }
+
+  PersonalityProvider(...args) {
```

Comment:
I'm confused by 142-147.  In the old code, we were setting `AffinityProvider` to point to the right class and then instantiated it. Here, we've wrapped the constructor with the same name, so we can call it? Why aren't we just saying something like `return new UserDomainAffinityProvider()` in line 139 and 133?

Reply:
It just makes it easier to stub these functions in a unit test, instead of stubbing an external jsm, I can stub *just* this function, return what I want, or check to see if it's called or not, without having to worry about external things. The contents of this function can also be tested to just return a thing.

It was a suggestion I got when I asked how the heck to stub remoteSettings.jsm, which is un stubable because of it's path.

I'm not a huge fan of it, but in the case of remote settings it is a pattern that solved a problem, and it is much easier to test for other jsms.